trieste/acquisition/function.py

```diff
@@ -676,3 +676,243 @@ def batch_ei(at: TensorType) -> TensorType:
+    when first preparing the acquisition function with no pending points. These estaimtes
+    are reused for all subsequent function calls.
+    """
+
+    def __init__(self, search_space: SearchSpace, grid_size: int = 500):
+        """
+        :param search_space: The global search space over which the optimisation is defined.
+        :param grid_size: Size of the grid over which the Lipschitz constant is estimated.
+            We recommend scaling this with search space dimension.
+        """
+        self._search_space = search_space
+        if grid_size <= 0:
+            raise ValueError(f"grid_size must be positive, got {grid_size}")
+        self._grid_size = grid_size
+
+        self._lipschitz_constant = None
+        self._eta = None
+
+    def prepare_acquisition_function(
+        self,
+        dataset: Dataset,
+        model: ProbabilisticModel,
+        pending_points: Optional[TensorType] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: The data from the observer.
+        :param model: The model over the specified ``dataset``.
+        :param pending_points: The points we penalize with respect to.
+        :return: The (log) expected improvement penalized with respect to the pending points.
+        :raise ValueError: if the first call does not have pending_points=None.
+        """
+        if len(dataset.query_points) == 0:
+            raise ValueError("Dataset must be populated.")
+
+        if (
+            pending_points is None
+        ):  # only compute penalization parameters once per optimization step
+            samples = self._search_space.sample(num_samples=self._grid_size)
+            samples = tf.concat([dataset.query_points, samples], 0)
+
+            def get_lipschitz_estimate(
+                sampled_points,
+            ) -> tf.Tensor:  # use max norm of posterior mean gradients
+                with tf.GradientTape() as g:
+                    g.watch(sampled_points)
+                    mean, _ = model.predict(sampled_points)
+                grads = g.gradient(mean, sampled_points)
+                grads_norm = tf.norm(grads, axis=1)
+                max_grads_norm = tf.reduce_max(grads_norm)
+                eta = tf.reduce_min(mean, axis=0)
+                return max_grads_norm, eta
+
+            lipschitz_constant, eta = get_lipschitz_estimate(samples)
+            if (
+                lipschitz_constant < 1e-5
+            ):  # threshold to improve numerical stability for 'flat' models
+                lipschitz_constant = 10
+
+            self._lipschitz_constant = lipschitz_constant
+            self._eta = eta
+
+        if self._lipschitz_constant is None:
+            raise ValueError("Local penalization must be first called with no pending_points.")
+
+        def log_acquisition(x: TensorType) -> TensorType:
+            return tf.math.log(expected_improvement(model, self._eta)(x))
```

Comment:
Is it important to return values in log space? otherwise I would suggest back-transforming to EI space for consistency with classical EI (here and below). EI value could be used for debugging (positivity) and to track down convergence. Alternatively add `Log` to the function name...